connect/src/spirc.rs

```diff
@@ -193,7 +196,13 @@ fn calc_logarithmic_volume(volume: u16) -> u16 {
     val
 }
 
-fn volume_to_mixer(volume: u16, linear_volume: bool) -> u16 {
+fn volume_to_mixer(volume: u16, linear_volume: bool, cache: Option<Cache>) -> u16 {
```

Comment:
This function shouldn't have side effects. Instead save_volume should be called by whoever also calls this method.

You could make a new private `set_volume` on Source, which updates the volume for all of `self.device`, `self.mixer` and `self.cache` and call that everytime.